### 1. Querying Data - SELECT, FROM, WHERE, ORDER BY, ASC, DESC

Queries are an essential tool in SQL. This workbook aims to give you the tools you need to be able to query data and obtain the information you want.

###### Keywords Covered: SELECT, FROM, PRAGMA, WHERE, ORDER BY, ASC, DESC

In [1]:
import sqlite3
import pandas as pd

# Connect to database
data_conn = sqlite3.connect("pokemon.db")

#### SELECT

Use: Specifying which columns you want to retrieve.

*Tip: Use * to select all columns. Multiple columns can be selected by seperating them with commas.*

#### FROM

Use: Specifies the table that you want to get the data from.

Below we will SELECT all columns from the pokemon table.

In [2]:
data = data_conn.execute("SELECT * FROM pokemon").fetchall()

print(f"Rows in data: {len(data)}")
print(data[0:3]) # Print first 3 rows.

Rows in data: 1025
[(1, 'Bulbasaur', 'grass', 'poison', 0.7, 6.9, 'While it is young, it uses the nutrients that are stored in the seed on its back in order to grow.', 'https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/official-artwork/1.png', 1500.0, 0, 0), (2, 'Ivysaur', 'grass', 'poison', 1.0, 13.0, 'Exposure to sunlight adds to its strength. Sunlight also makes the bud on its back grow larger.', 'https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/official-artwork/2.png', 1500.0, 0, 0), (3, 'Venusaur', 'grass', 'poison', 2.0, 100.0, 'A bewitching aroma wafts from its flower. The fragrance becalms those engaged in a battle.', 'https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/official-artwork/3.png', 1500.0, 0, 0)]


In [3]:
# Display in a cleaner format by putting the data into a dataframe.
df = pd.DataFrame(data)
df.head(3)

,0,1,2,3,4,5,6,7,8,9,10
0,1,Bulbasaur,grass,poison,0.7,6.9,"While it is young, it uses the nutrients that ...",https://raw.githubusercontent.com/PokeAPI/spri...,1500.0,0,0
1,2,Ivysaur,grass,poison,1.0,13.0,Exposure to sunlight adds to its strength. Sun...,https://raw.githubusercontent.com/PokeAPI/spri...,1500.0,0,0
2,3,Venusaur,grass,poison,2.0,100.0,A bewitching aroma wafts from its flower. The ...,https://raw.githubusercontent.com/PokeAPI/spri...,1500.0,0,0


This is a large amount of data, 1025 rows, to view all at once. We need to filter it to selectively to view relevant parts. To do this we first need to view the column names in the pokemon table using the PRAGMA keyword to know what we can filter by.

#### PRAGMA

Use: Query internal database information like column names.

*Note: This keyword is unique to SQLite.*

In [4]:
# Retrieve column data.
cols = data_conn.execute("PRAGMA table_info(pokemon)").fetchall()

for col in cols:
    print(col)

(0, 'id', 'INTEGER', 0, None, 1)
(1, 'name', 'TEXT', 1, None, 0)
(2, 'type1', 'TEXT', 1, None, 0)
(3, 'type2', 'TEXT', 0, None, 0)
(4, 'height', 'REAL', 1, None, 0)
(5, 'weight', 'REAL', 1, None, 0)
(6, 'dex_entry', 'TEXT', 1, "''", 0)
(7, 'image_url', 'TEXT', 1, None, 0)
(8, 'elo', 'REAL', 1, '1500', 0)
(9, 'wins', 'INTEGER', 1, '0', 0)
(10, 'losses', 'INTEGER', 1, '0', 0)


In [5]:
# This can also be used to list the tables in the database.
tables = data_conn.execute("PRAGMA table_list").fetchall()

for table in tables:
    print(table)

('main', 'votes', 'table', 5, 0, 0)
('main', 'pokemon', 'table', 11, 0, 0)
('main', 'sqlite_sequence', 'table', 2, 0, 0)
('main', 'users', 'table', 4, 0, 0)
('main', 'sqlite_schema', 'table', 5, 0, 0)
('temp', 'sqlite_temp_schema', 'table', 5, 0, 0)


'sqlite_schema', sqlite_temp_schema' and 'sqlite_sequence are the default SQLite tables that contain the structure, temporary objects and autoincrement values. pokemon, votes and users are tables that store data.

### Filtering Queries

Now we know the column names for the pokemon table are: 'id', 'name', 'type1', 'type2', 'height', 'weight', 'dex_entry', 'image_url', 'elo', 'wins' and 'losses'.

Say we want to know how many type1 Water Pokemon there are. We can do this using the WHERE keyword.

#### WHERE

Use: Filtering queries by a specified constraint.

In [6]:
water_pkmn = data_conn.execute("SELECT * FROM pokemon WHERE type1 = 'water'").fetchall()

print(f"Water Pokemon in data: {len(water_pkmn)}")

Water Pokemon in data: 134


WHERE can be used with conjunctions like AND to filter in more than one way.

#### AND

Use: Filtering records based on more than one condition.

For instance, if we wanted know not just how many water Pokeone there were in total, but how many there are in the original 151 Pokemon.

In [7]:
water_pkmn = data_conn.execute("SELECT * FROM pokemon WHERE type1 = 'water' and id < 152").fetchall()

print(f"Water Pokemon in first 151 Pokemon: {len(water_pkmn)}")

Water Pokemon in first 151 Pokemon: 28


The next Keyword we could use in a query is ORDER BY.

#### ORDER BY

Use: Ordering data by a certain column in ascending or descending order

#### ASC

Use: Specifying the data to be sorted in ascending order (smallest to largest).

#### DESC

Use: Specifying the data to be sorted in descending order (largest to smallest).

In the below example, we will list the Pokemon in reverse order by listing by their id in descending order.

In [8]:
desc_pkmn = data_conn.execute("SELECT * FROM pokemon ORDER BY id DESC").fetchall()
pd.DataFrame(desc_pkmn).head(3)

,0,1,2,3,4,5,6,7,8,9,10
0,1025,Pecharunt,poison,ghost,0.3,0.3,Its peach-shaped shell serves as storage for a...,https://raw.githubusercontent.com/PokeAPI/spri...,1500.0,0,0
1,1024,Terapagos,normal,NaN,0.2,6.5,It’s thought that this Pokémon lived in ancien...,https://raw.githubusercontent.com/PokeAPI/spri...,1500.0,0,0
2,1023,Iron Crown,steel,psychic,1.6,156.0,There was supposedly an incident in which it l...,https://raw.githubusercontent.com/PokeAPI/spri...,1500.0,0,0


Lastly, some notes on the fetchall() function we have been using. As well as this there are two other variants, fetchone() that returns one row, or fetchmany(), which returns a list of tuples if you want results in batches.


|Method | What it returns | When to use |
| --- | --- | --- |
|.fetchall() | List of tuples (all rows) | When you want all results at once |
|.fetchone() |Single tuple (or None) | When you expect only 1 row |
|.fetchmany(size) | List of tuples (up to size) | When you want results in batches |

In [9]:
# As shown, using fetchone() will only fetch the first result of a query.
one_pkmn = data_conn.execute("SELECT * FROM pokemon").fetchone()
print(one_pkmn)

(1, 'Bulbasaur', 'grass', 'poison', 0.7, 6.9, 'While it is young, it uses the nutrients that are stored in the seed on its back in order to grow.', 'https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/official-artwork/1.png', 1500.0, 0, 0)


In [10]:
# Using fetchmany() will fetch the number of rows specified. In this case the first three electric Pokemon.
three_pkmn = data_conn.execute("SELECT * FROM pokemon WHERE type1 = 'electric'").fetchmany(3)
pd.DataFrame(three_pkmn).head(3)

,0,1,2,3,4,5,6,7,8,9,10
0,25,Pikachu,electric,NaN,0.4,6.0,Possesses cheek sacs in which it stores electr...,https://raw.githubusercontent.com/PokeAPI/spri...,1500.0,0,0
1,26,Raichu,electric,NaN,0.8,30.0,It can discharge bursts of electricity exceedi...,https://raw.githubusercontent.com/PokeAPI/spri...,1500.0,0,0
2,81,Magnemite,electric,steel,0.3,6.0,A bizarre Pokémon with but a single eye embedd...,https://raw.githubusercontent.com/PokeAPI/spri...,1500.0,0,0
